src/_nebari/upgrade.py

```diff
@@ -1297,11 +1298,34 @@ def _version_specific_upgrade(
 
             urllib3.disable_warnings()
 
+            keycloak_username = os.environ.get("KEYCLOAK_ADMIN_USERNAME", "root")
+            keycloak_password = os.environ.get(
+                "KEYCLOAK_ADMIN_PASSWORD",
+                config["security"]["keycloak"]["initial_root_password"],
+            )
+
             keycloak_admin = get_keycloak_admin(
                 server_url=f"https://{config['domain']}/auth/",
-                username="root",
-                password=config["security"]["keycloak"]["initial_root_password"],
+                username=keycloak_username,
+                password=keycloak_password,
             )
+
+            try:
+                # Quick test to connect to Keycloak
+                keycloak_admin.get_server_info()
+            except ValueError as e:
+                if "invalid_grant" in str(e):
+                    print(
+                        "[red bold]Failed to connect to the Keycloak server.[/red bold]\n"
+                        "[yellow]This may occur if the default admin credentials have been changed for security reasons.[/yellow]\n"
```

Comment:
I feel like this makes it sound like the user did something wrong or optional when in reality changing the security credentials should be the default behaviors. I'll suggest an edit.